### PR TITLE
[Music]Fetching local art for multi-folder albums (disc sets)

### PR DESCRIPTION
**Pull request by DaveTBlake**
Fix a long standing problem fetching local art when the album is split across multiple subfolders e.g. disc sets. Generally improve handling of art for disc sets as a small _initial step_ in enhancing disc set support. Make some other small improvements to fetching local artist art too.
___
Users  have long reported that sometimes images from the album folder get used as artist art, and not getting the album art they expected. See https://forum.kodi.tv/showthread.php?tid=329388&pid=2725958#pid2725958 for one recent example.  One cause at least was the way scanning treated music files in this kind of layout:
```
mymusic/artist1/album1/CD1/track01.mp3
mymusic/artist1/album1/CD1/track02.mp3
...
mymusic/artist1/album1/CD2/track01.mp3
mymusic/artist1/album1/CD2/track02.mp3
...
```
It was assuming that the parent and grand-parent folders of the music files would be the album and artist folders respectively, and they are not.  Each subfolder was scanned independantly of the other and treated as the album folder, so the album art was taken from last subfolder or the songs in that subfolder only.

The solution was to have a “fetch local art” phase for the processed albums and artists once file scan completed. This made it possible  make use of art from the subfolders, when they contain music files with the same DISKNUMBER tag, as fallback song thumbs. So for example if `mymusic/artist1/album1/CD1/folder.jpg` exists, and all the music files have the same album and same embedded cover art, then it will appear when songs from that disc are listed or played.  

Fetching local art is part of both initial scanning, and any subsequent scraping from remote sites or NFO files (on "Query for all" or "Refresh").  When artist art was missing, and no infomation was fetched from remote sites, the process did not check for local art. It meant that when a user added local art for such a lesser known artist they either had to set it manually, add and NFO file too or drop and re-add the music files for that artist. This is also fixed, and checking locally for missing art done as part of scraping.

To aid manual art selection, especially for disc sets, the Song Info Dialog now offers album folder as the `*itemfolder` when browsing for art, in preference to the parent folder of the song (not always the same).

Tested myself using all combinations of albums with multiple subfolders and art, but will put up a test build and invite more user testing.

### CODE REVIEW

**Comment by DaveTBlake** [comment]
Test build available on the mirrors, and user feedback positive so far. Meanwhile anyone care to code review?

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -5461,70 +5461,99 @@ int CMusicDatabase::GetSongsCount(const Filter &filter)
...
 
-bool CMusicDatabase::GetAlbumPath(int idAlbum, std::string &basePath)
+bool CMusicDatabase::GetAlbumPath(int idAlbum, std::string& basePath)
 {
+  basePath.clear();
+  std::vector<std::pair<std::string, int>> paths;
+  if (!GetAlbumPaths(idAlbum, paths))
+    return false;
+
+  for (auto pathpair : paths)
```
Comment (Rechi) [inline xbmc/music/MusicDatabase.cpp]: `const auto& pathpair`

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 2]
```diff
@@ -1177,6 +1178,143 @@ void CMusicInfoScanner::FindArtForAlbums(VECALBUMS &albums, const std::string &p
...
+    It has already been during set by FindArtForAlbums() to either the art of
+    the last subfolder processed (if there is any), or to the first song in
+    that subfolder with embedded art (if there is any).
+    Not when songs from different albums are in one folder, no paths are returned.
+    */
+
+    std::vector<std::pair<std::string, int>> paths;
+    m_musicDatabase.GetAlbumPaths(albumId, paths);
+    // Get album path, the common path when more than one 
+    for (auto pathpair : paths)
```
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: `const auto& pathpair`

File: xbmc/music/infoscanner/MusicInfoScanner.cpp [hunk 3]
```diff
@@ -1177,6 +1178,143 @@ void CMusicInfoScanner::FindArtForAlbums(VECALBUMS &albums, const std::string &p
...
+      std::string albumArt = albumItem.GetUserMusicThumb(true);
+
+      /*
+      When we have a true disc set - subfolders AND songs tagged with same
+      unique discnumber in in each subfolder - save the disc cover art, and if
+      we don't have album folder art then use the first disc in set rather
+      than the last processed.
+      */
+      CMusicThumbLoader loader;
+      for (auto pathpair : paths)
```
Comment (Rechi) [inline xbmc/music/infoscanner/MusicInfoScanner.cpp]: `const auto& pathpair`

**Comment by DaveTBlake** [comment]
If there are no objections I propose to merge this on Monday. Thanks for code check @Rechi

**Comment by garbear** [comment]
Code looks good. A suggestion I have is to break up the long documentation and disperse it throughout the code. Better yet, you can break up the long functions into smaller ones with self-describing names, so much of the documentation in the source can disappear. The ultimate documentation bliss is when the code is so clear, documentation exists only in the header, not the source.

**Comment by DaveTBlake** [comment]
@garbear I agree in general about making code as readable and self explanatory as possible. However sometimes an overview of the design is needed too, and the music library scanning and scraping process is one of those. Kodi does not have design documentation, and as well as putting a trail of explanations in both Github and on the forum, something immediately in the code is useful. At least I will find it useful if I come back to it years later, and I believe those that come after me will too.  It certainly does no harm.

**Pull request merged**